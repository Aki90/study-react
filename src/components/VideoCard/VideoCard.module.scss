@import '../../styles/mixins/line-clamp';

.VideoCard {
  display: inline-flex;
  flex-direction: column;
  width: 270px;
  height: 300px;
  padding: 10px;
  box-sizing: border-box;

  &:hover {
    cursor: pointer;
  }

  .title {
    @include line-clamp(1);
  }

  .thumbnail {
    display: block;
    width: 100%;
    height: 50%;
    object-fit: cover;
    border-radius: 10px;
  }
}
