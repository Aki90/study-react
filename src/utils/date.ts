import { TDate, format, register } from 'timeago.js';
import koLocale from 'timeago.js/lib/lang/ko';

register('ko', koLocale);

export function formatAgo(date: TDate, lang = 'ko') {
  return format(date, lang);
}
