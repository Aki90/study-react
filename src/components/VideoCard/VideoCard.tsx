/* eslint-disable @typescript-eslint/no-explicit-any */
import { formatAgo } from '@/utils/date';
import { useNavigate } from 'react-router';
import styles from './VideoCard.module.scss';

export default function VideoCard({
  video,
  videoId,
}: {
  video: any;
  videoId: string;
}) {
  const navigate = useNavigate();
  const { title, thumbnails, channelTitle, publishedAt } = video.snippet;

  const handleNavigate = (): void => {
    navigate(`/videos/watch/${videoId}`, { state: { video } });
  };

  return (
    <li className={styles.VideoCard} onClick={handleNavigate}>
      <img
        className={styles.thumbnail}
        src={thumbnails.medium.url}
        alt={title}
      />
      <div>
        <p className={styles.title}>{title}</p>
        <p>{channelTitle}</p>
        <p>{formatAgo(publishedAt)}</p>
      </div>
    </li>
  );
}
