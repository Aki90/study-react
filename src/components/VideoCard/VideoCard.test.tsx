/* eslint-disable testing-library/no-unnecessary-act */
import { screen, render, act } from '@testing-library/react';
import { withRouter } from '@/tests/utils';
import VideoCard from './VideoCard';

import { Route, useLocation } from 'react-router-dom';
import userEvent from '@testing-library/user-event';

describe('<VideoCard />', () => {
  const video = {
    id: 'id',
    snippet: {
      title: 'title',
      channelId: '1',
      channelTitle: 'channelTitle',
      publishedAt: new Date(),
      thumbnails: {
        medium: {
          url: 'http://image/',
        },
      },
    },
  };

  test('renders correctly', async () => {
    const { container } = render(
      withRouter(<Route path="/" element={<VideoCard video={video} />} />),
    );

    const image = screen.getByRole('img') as HTMLImageElement;
    const { thumbnails, title } = video.snippet;
    expect(image.src).toBe(thumbnails.medium.url);
    expect(image.alt).toBe(title);
    expect(screen.getByText(title)).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  test('video item 클릭 시, state 를 담아 video detail 페이지로 이동해야 한다.', async () => {
    function LocationStateDisplay() {
      return <pre>{JSON.stringify(useLocation().state)}</pre>;
    }

    render(
      withRouter(
        <>
          <Route path="/" element={<VideoCard video={video} />} />
          <Route
            path={`/videos/watch/${video.id}`}
            element={<LocationStateDisplay />}
          />
        </>,
      ),
    );

    const card = screen.getByRole('listitem');

    await act(async () => {
      userEvent.click(card);
    });

    expect(screen.getByText(JSON.stringify({ video }))).toBeInTheDocument();
  });
});
