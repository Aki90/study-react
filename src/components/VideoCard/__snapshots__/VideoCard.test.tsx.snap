// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VideoCard /> renders correctly 1`] = `
<div>
  <li
    class="VideoCard"
  >
    <img
      alt="title"
      class="thumbnail"
      src="http://image/"
    />
    <div>
      <p
        class="title"
      >
        title
      </p>
      <p>
        channelTitle
      </p>
      <p>
        방금
      </p>
    </div>
  </li>
</div>
`;
